avm2: Refactor object methods to do multiname lookups

kmeisthax opened the pull request:
This PR includes a number of refactors necessary for future AVM2 work. Since these are cross-cutting concerns and imply a lot of code churn, I'm packaging them separately from the work that will benefit from it.

 * ~~Multinames can now hold an `Object` directly.~~ Nope.
 * `TObject` methods such as `get_property`, `set_property`, and `init_property` now take a `Multiname` and do the name resolution internally.
 * `call_property` and `construct_prop` are now `TObject` methods, and existing code has been changed to use them.
 * Setting dynamic properties outside of the public namespace is no longer permitted, and dynamic properties on sealed classes is now forbidden entirely. Instance slots have been added for impacted native class implementations.
 * Prototypes of classes are no longer instances of that class (required by the prior rule change)
   * This is required by the fact that we currently do not allow calls to unbound methods through the prototype chain, which AVM2 actually does allow. I cannot implement that until both this PR and #5325 land; and in practice this only matters for ES3 legacy support.

This will enable ~~implementation of `Dictionary`, as well as~~ optimizations to method lookup, such as not instantiating bound-receiver copies of methods for each call.

EmperorBale commented:
Is there a reason multinames are restricted to holding only a String or Object? Could it maybe just hold a Value instead? Holding a Value may be helpful for indexing Arrays, because currently we need to parse the string to an integer.

adrian17 commented:
> Multinames can now hold an Object directly.
> TObject methods such as get_property, set_property, and init_property now take a Multiname and do the name resolution internally.

Did you see our earlier thoughts on discord, partially based on comparing with avmplus (check out `INSTR(getproperty)`)? Changing `op_{get,set,delete}_property` opcodes to separate cases for names and objects (and possibly integers too) and pass them to separate TObject methods might be a better choice in the long run.

kmeisthax commented:
I did consider separate `get_property_by_string`/`_by_object` methods, but I figured that it would add more complexity to TObject implementation purely for one particular class. Most classes are going to use standard multiname resolution (i.e. coerce local name to string, check all namespaces for the key, and yield a normal `QName`). The future `Dictionary` PR can change this by instead inspecting the type of the local name to select what part of the object to store the value in.

I'll consider replacing `LocalName` with `Value` to support array-like objects. It should at least let us skip the parsing step for known numbers; though for string properties we will still need to run a parse - AFAIK [1] and ["1"] are the same name.

adrian17 commented:
Hmm, personally I think the current approach also adds some extra complexity, like nesting an Object in a Multiname - I was recently looking for a way to simplify it instead :D This also feels kinda hack-y to me.

> but I figured that it would add more complexity to TObject implementation purely for one particular class

We can alternatively special-case for Dictionary directly in the `op_{get,set,delete}property` instead (if we're 100% sure we don't need a special case for an Object anywhere else) - this way we don't need to add extra methods and complexity to all `TObject` implementations.

And personally I still like separate path for numeric indices will be nice - it'll be useful not only for array/vector-like objects, but also for low-overhead internal object indexing.

kmeisthax commented:
I rebased and ripped out all of the code relating to `Multiname` holding an object.

For the record, I generally don't consider "avmplus does it this way" on it's own to be a compelling argument. We have a duty, legally speaking, to not slavishly copy MPL code. (Yes, C++ and Rust aren't 1:1 copies, but a court probably would still consider that a derivative work.)

However, I'm not opposed to doing what avmplus does, if there's other arguments for doing it that way. And, after playing around with `Dictionary` and `Proxy` a bit, I'm pretty sure that we actually will need to special-case the former in the actual AVM2 opcode implementations. The thing that convinced me was that `Proxy` does *not* provide a way to get an object reference from the name it gets - there's no way to write `Dictionary` with just `Proxy`. Furthermore, `Dictionary` does not handle object-key lookups on other namespaces - those get coerced to string as part of a standard property lookup.

Ergo, the only benefit from implementing it my proposed way is that `op_xyz_property` and `Object.xyz_property` are in one-to-one correspondence. I don't think that's a good enough argument, so I removed the relevant commits from this PR. `Dictionary` will now be special-cased instead. This also means we could add more fast paths for other types of lookups on array/vector-structured objects without necessarily making them into `Multiname`s.

adrian17 commented on core/src/avm2/activation.rs:
```diff
@@ -1086,8 +1086,8 @@ impl<'a, 'gc, 'gc_context> Activation<'a, 'gc, 'gc_context> {
             .ok_or_else(|| format!("Object method {} does not exist", index.0).into());
-        let base_proto = receiver.proto();
-        let value = function?.call(Some(receiver), &args, self, base_proto)?;
+        let superclass_object = receiver.instance_of();
```
General note: the naming over here is confusing. We get an object's class with `instance_of`, store it in variable `superclass_object` (which in our naming is an entirely different thing), which is then passed into argument and Activation field named `subclass_object`.

kmeisthax replied:
Right. So, `superclass_object` is supposed to represent the class that defined a particular method, so that subsequent calls to `super` progress up the class chain correctly. `super` is also supposed to work on bound methods, which is why we currently pull `instance_of` in a lot of places where we have a bare method. I suspect the correct thing to do would be to actually eliminate the `superclass_object` parameter from `call` entirely and just have methods bind to the class they were defined on (in which case we could rename this to `defining_class`).

However, I'm not going to do that until I implement lazy method binding, as I suspect that feature will interact with bound defining classes in an unintuitive way.

adrian17 reviewed (approved):
LGTM, aside from naming note above.

Long-term I'm feeling we'll need something to make the multinames cheaper or avoid them for internal calls, but that's a future concern.
